### ISSUE TITLE
Querystring schema deserializes request header?

### ISSUE THREAD

**Issue by jenstroeger** [Cornices/cornice#449]
I am not sure if this is a problem on the [Cornice](https://github.com/Cornices/cornice) side, a [Colander](https://github.com/Pylons/colander) schema problem, or actually a bug in my code because I didn’t understand how to put the pieces together 🤓

I've got the following schema hierarchy:
```python
class StrictMappingSchema(colander.MappingSchema):                                                  
    def schema_type(self, **kw):                                                                    
        return colander.Mapping(unknown="raise")                                                    

class DataSchema(StrictMappingSchema):
    foo = colander.SchemaNode(…)
    bar = colander.SchemaNode(…)

class QuerystringDataSchema(StrictMappingSchema):
    querystring = DataSchema()
```
The view function:
```python
@awesome_service.get( 
   content_type="application/json",
    accept="application/json", 
    permission="awesome_get",
    schema=QuerystringDataSchema,
    validators=(
        colander_validator,
        )
    )
def awesome_get(request):
    …
```
Sending a request with valid or invalid querystring data _always_ causes a `400 Bad Request` where the error description contains a dictionary of the (entire) request header:
```json
{
    "errors": [
        {
            "description": "Unrecognized keys in mapping: \"{'header': {'Accept': 'application/json, */*', 'User-Agent': 'HTTPie/0.9.9', 'Authorization': 'Bearer …', 'Connection': 'keep-alive', 'Host': 'localhost:6543', 'Content-Type': 'application/json', 'Accept-Encoding': 'gzip, deflate'}, 'cookies': {}, 'url': 'http://localhost:6543/api/…', 'method': 'GET', 'path': {}, 'body': {}}\"",
            "location": "",
            "name": ""
        }
    ],
    "status": "error"
}
```
This can be “fixed” in two ways:
- Set `unknown="ignore"` in the `StrictMappingSchema`, thus defeating the purpose of that schema’s existence.
- Derive `QuerystringDataSchema` from `colander.MappingSchema` instead of `StrictMappingSchema`. Why would that make a difference?

**Comment by leplatrem** [comment]
Thanks for this detailed feedback ;) 

So, by default `colander` ignores unknown keys. If you use `colander.MappingSchema` it is the same as using `ignore`.

The request schema will receive data like `headers`, `body`, `querystring`, `path`, `cookies`.  If you use a strict schema that raises when something is unknown, you should define a sub-schema for each of those.

I believe that what you want is to validate the querystring and ignore the rest. In that case, it would just be:

```py
class RequestDataSchema(colander.MappingSchema):
    querystring = DataSchema()
```

Let us know ;)

**Comment by jenstroeger** [comment]
That was the fix I used, actually, but I wanted to make sure it’s the recommended way of going about this.

I also think that my question above is related to https://github.com/Cornices/cornice/issues/430 which I filed a while ago. Considering that there is a `colander_body_validator` available, would it not make sense to add explicit validators for each of the data blobs you mention, `body`, `querystring`, `cookies` at least?

Shall I close this issue then?
